native/src/fairseq2n/data/sample_data_source.cc

```diff
@@ -170,17 +181,49 @@ sample_data_source::next_in_pipeline(std::size_t pipeline_idx)
     if (!maybe_example) {
         is_epoch_done_[pipeline_idx] = true;
 
-        pipeline.reset();
+        if (allow_repeats_) {
+            pipeline.reset();
 
-        // Circle back to the first example.
-        maybe_example = pipeline.next();
-        if (!maybe_example)
-            throw_data_pipeline_error(/*maybe_example=*/std::nullopt, /*recoverable=*/false,
-                "The data pipeline at index {} is empty and cannot be sampled.", pipeline_idx);
+            // Circle back to the first example.
+            maybe_example = pipeline.next();
+            if (!maybe_example)
+                throw_data_pipeline_error(/*maybe_example=*/std::nullopt, /*recoverable=*/false,
+                    "The data pipeline at index {} is empty and cannot be sampled.", pipeline_idx);
+        } else
+            block(pipeline_idx);
     } else if (pipeline.finitude_type() == data_source_finitude_type::pseudo_infinite)
         is_epoch_done_[pipeline_idx] = true;
 
-    return std::move(*maybe_example);
+    return maybe_example;
+}
+
+void
+sample_data_source::block(std::size_t idx)
+{
+    float32 weight = weight_cumsums_[idx];
+    if (idx > 0) {
+        weight -= weight_cumsums_[idx - 1];
+        weight_cumsums_[idx] = weight_cumsums_[idx - 1];
+    } else {
+        weight_cumsums_[idx] = 0.0F;
+    }
+    for (std::size_t i = idx + 1; i < weight_cumsums_.size(); ++i) {
+        weight_cumsums_[i] -= weight;
+    }
+
+    float32 sum = weight_cumsums_.back();
+
+    if (!are_close(sum, 1.0F)) {
+        for (float32 &s : weight_cumsums_)
+            s /= sum;
+    }
+}
+
+void
+sample_data_source::sum_weights()
```

Comment:
Is this function used somewhere? I cannot see any references to it in this PR.

Reply:
My mistake.